### PR TITLE
feat(pt_expt): add DeepSpin support for pt_expt backend

### PR DESCRIPTION
**Pull request by wanghan-iapcm**
## Summary
- Add DeepSpin (spin model) support for the pt_expt (.pt2) backend: C++ API, Python inference, LAMMPS pair style, and model deviation
- Fix `add_mag` bug in dpmodel `SpinModel`: atomic virial used `add_mag=False` while PT backend used `add_mag=True`, causing inconsistent per-atom virial between .pth and .pt2
- Add `"type": "spin_ener"` to spin model serialization for correct deserialization dispatch across all backends
- Replace committed `.pth` model file with canonical `.yaml` (dpmodel serialization); `.pth` and `.pt2` are now regenerated from `.yaml` via `convert_backend` at test time

## Test plan
- [x] C++ spin tests: 38 passed (12 TF-only skipped)
- [x] LAMMPS .pt2 spin tests: 7 passed (PBC + NoPBC + model_devi)
- [x] LAMMPS .pth spin tests: 3 passed (single-model; model_devi needs TF .pb)
- [x] .pth vs .pt2 agreement: all quantities diff ~1e-15
- [ ] CI: full test suite

## Summary by CodeRabbit

* **New Features**
  * Spin-capable models: public queries for model spin presence and per-type spin usage; PyTorch-exportable spin backend and new export/run backend for spin models; serialized spin model discriminator added.
  * Output pipeline extended to include magnetic derivatives, magnetic virial counterparts, and per-atom magnetic masks.
* **Tests**
  * Added extensive unit and regression tests, new test suites and generators for spin models, updated golden reference data, and model-generation scripts.

### CODE REVIEW

File: source/api_cc/src/DeepSpinPTExpt.cc [hunk 1]
```diff
@@ -0,0 +1,800 @@
...
+      max_x = std::max(max_x, coord_d[ii * 3 + 0]);
+      min_y = std::min(min_y, coord_d[ii * 3 + 1]);
+      max_y = std::max(max_y, coord_d[ii * 3 + 1]);
+      min_z = std::min(min_z, coord_d[ii * 3 + 2]);
+      max_z = std::max(max_z, coord_d[ii * 3 + 2]);
+    }
+    box_d.resize(9, 0.0);
+    box_d[0] = (max_x - min_x) + 2.0 * rcut;
+    box_d[4] = (max_y - min_y) + 2.0 * rcut;
+    box_d[8] = (max_z - min_z) + 2.0 * rcut;
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/DeepSpinPTExpt.cc:522]: Fixed in 703d9939 — added coordinate shift in the NoPBC path of DeepSpinPTExpt::compute, matching DeepPotPTExpt.

File: deepmd/pt_expt/utils/serialization.py [hunk 2]
```diff
@@ -317,64 +372,105 @@ def _trace_and_export(
 
     target_device = _env.DEVICE
 
+    # Detect spin model
+    is_spin = data["model"].get("type") == "spin_ener"
+
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/utils/serialization.py:387]: Not applicable — `"type": "spin_ener"` was added in this PR, and there are no pre-existing .pt2 spin models (spin .pt2 support is introduced by this PR). All spin model serialization now includes the `type` key.

File: deepmd/pt_expt/infer/deep_eval.py [hunk 3]
```diff
@@ -695,6 +743,143 @@ def _eval_model(
...
+
+        if fparam is not None:
+            fparam_t = torch.tensor(
+                fparam.reshape(nframes, self.get_dim_fparam()),
+                dtype=torch.float64,
+                device=DEVICE,
+            )
+        else:
+            fparam_t = None
+
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/infer/deep_eval.py:846]: Fixed in 6e048efa — removed the `_is_pt2` guard so default_fparam fallback works for both .pt2 and .pte in both spin and non-spin paths. Added TestSpinDefaultFparam test covering both formats.

File: deepmd/pd/infer/deep_eval.py [hunk 4]
```diff
@@ -297,6 +297,13 @@ def get_has_spin(self) -> bool:
         """Check if the model has spin atom types."""
         return self._has_spin
 
+    def get_use_spin(self) -> list[bool]:
+        """Get the per-type spin usage of this model."""
+        if self._has_spin:
+            model = self.dp.model["Default"]
+            return model.spin.use_spin.tolist()
+        return []
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pd/infer/deep_eval.py:305]: Not applicable — PD backend doesn't support spin models (throws NotImplementedError). This is out of scope for this PR.

File: deepmd/pt_expt/infer/deep_eval.py [hunk 5]
```diff
@@ -283,9 +323,16 @@ def eval(
...
+        spins = kwargs.get("spin")
+        if spins is not None:
+            spins = np.array(spins)
+            out = self._eval_func(self._eval_model_spin, numb_test, natoms)(
+                coords, cells, atom_types, spins, fparam, aparam, request_defs
+            )
+        else:
+            out = self._eval_func(self._eval_model, numb_test, natoms)(
+                coords, cells, atom_types, fparam, aparam, request_defs
+            )
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/infer/deep_eval.py:345]: The PT backend uses the exact same pattern — dispatch based on whether `spin` kwarg is present. Consistent with existing behavior.

File: source/api_cc/src/commonPTExpt.h [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+};
+
+class JsonParser {
+ public:
+  explicit JsonParser(const std::string& s) : s_(s), pos_(0) {}
+  JsonValue parse() {
+    skip_ws();
+    auto val = parse_value();
+    return val;
+  }
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/commonPTExpt.h:59]: The JSON is always generated by `json.dumps()` in Python serialization code, so it will always be well-formed. Not worth the added complexity.

File: source/api_cc/src/commonPTExpt.h [hunk 7]
```diff
@@ -0,0 +1,532 @@
...
+           (static_cast<uint32_t>(
+                static_cast<unsigned char>(content[offset + 1]))
+            << 8) |
+           (static_cast<uint32_t>(
+                static_cast<unsigned char>(content[offset + 2]))
+            << 16) |
+           (static_cast<uint32_t>(
+                static_cast<unsigned char>(content[offset + 3]))
+            << 24);
+  };
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/commonPTExpt.h:303]: The .pt2 files are produced by `torch.export` and Python's `zipfile` module, so they will always be well-formed. The parser already has a bounds check at line 348.

File: source/api_cc/src/commonPTExpt.h [hunk 8]
```diff
@@ -0,0 +1,532 @@
...
+        match = true;
+      }
+    }
+    if (match) {
+      uint16_t local_name_len = read_u16(local_header_offset + 26);
+      uint16_t local_extra_len = read_u16(local_header_offset + 28);
+      size_t data_offset =
+          local_header_offset + 30 + local_name_len + local_extra_len;
+      return content.substr(data_offset, uncompressed_size);
+    }
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/commonPTExpt.h:431]: PyTorch's `torch.export` + Python's `zipfile` use STORED (compression method 0) for all entries. No compressed entries will be encountered in practice.

File: source/api_cc/src/DeepSpin.cc [hunk 9]
```diff
@@ -723,3 +739,10 @@ template void DeepSpinModelDevi::compute<float>(
     const int& ago,
     const std::vector<float>& fparam,
     const std::vector<float>& aparam);
+
+std::vector<bool> DeepSpinModelDevi::get_use_spin() const {
+  if (!dps.empty()) {
+    return dps[0]->get_use_spin();
+  }
+  return {};
+}
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/DeepSpin.cc:748]: Standard pattern in the codebase — `DeepPotModelDevi` uses the same first-model approach. All models in a model deviation run must have the same use_spin config.

File: source/api_cc/src/DeepSpinPTExpt.cc [hunk 10]
```diff
@@ -0,0 +1,800 @@
...
+
+  if (atomic) {
+    // atom_energy: energy (nf, nloc, 1)
+    torch::Tensor atom_energy_tensor =
+        output_map["energy"].view({-1}).to(floatType);
+    torch::Tensor cpu_atom_energy_ = atom_energy_tensor.to(torch::kCPU);
+    atom_energy.assign(
+        cpu_atom_energy_.data_ptr<VALUETYPE>(),
+        cpu_atom_energy_.data_ptr<VALUETYPE>() + cpu_atom_energy_.numel());
+
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/DeepSpinPTExpt.cc:681]: False positive — `output_map["energy"]` IS the per-atom energy (shape nf, nloc, 1) from `forward_lower`. The total energy is `output_map["energy_redu"]`. DeepPotPTExpt uses the same pattern (lines 369, 619).

File: source/api_cc/tests/test_deeppot_dpa_ptexpt_spin.cc [hunk 11]
```diff
@@ -0,0 +1,507 @@
...
+    std::string model_path = "../../tests/infer/deeppot_dpa_spin.pt2";
+    {
+      std::ifstream f(model_path);
+      if (!f.good()) {
+        GTEST_SKIP() << "Skipping: " << model_path << " not found.";
+      }
+    }
+#ifndef BUILD_PYTORCH
+    GTEST_SKIP() << "Skip because PyTorch support is not enabled.";
+#endif
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/tests/test_deeppot_dpa_ptexpt_spin.cc:57]: Not necessary — if BUILD_PT_EXPT_SPIN is 0, DeepSpin::init() throws at runtime. The test also checks if the model file exists (line 48-53). Works correctly as-is.

File: source/lmp/tests/test_lammps_spin_nopbc_pt2.py [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+)
+
+pb_file = (
+    Path(__file__).parent.parent.parent / "tests" / "infer" / "deeppot_dpa_spin.pt2"
+)
+pb_file2 = (
+    Path(__file__).parent.parent.parent / "tests" / "infer" / "deeppot_dpa_spin_md1.pt2"
+)
+data_file = Path(__file__).parent / "data.lmp"
+md_file = Path(__file__).parent / "md.out"
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/lmp/tests/test_lammps_spin_nopbc_pt2.py:23]: Standard pattern used across all LAMMPS test files in the repo (test_lammps_spin_pt.py, test_lammps_pt2.py, etc). Changing this file would be inconsistent.

File: source/lmp/tests/test_lammps_spin_pt2.py [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+def setup_module() -> None:
+    if os.environ.get("ENABLE_PYTORCH", "1") != "1":
+        pytest.skip(
+            "Skip test because PyTorch support is not enabled.",
+        )
+    write_lmp_data_spin(box, coord, spin, type_NiO, data_file)
+
+
+def teardown_module() -> None:
+    os.remove(data_file)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/lmp/tests/test_lammps_spin_pt2.py:142]: Follows the existing teardown pattern in test_lammps_spin_pt.py (line 194). md_file is only created when model devi tests run.

File: source/tests/infer/gen_spin_model_devi.py [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+    lmp_box = np.array(
+        [13.0, 0.0, 0.0, 0.0, 13.0, 0.0, 0.0, 0.0, 13.0], dtype=np.float64
+    )
+
+    print("\n// ---- LAMMPS 4-atom system (PBC) ----")  # noqa: T201
+    for idx, pt2_path in enumerate(pt2_paths):
+        dp = DeepPot(pt2_path)
+        e, f, v, ae, av, fm, _ = dp.eval(
+            lmp_coord, lmp_box, lmp_atype, atomic=True, spin=lmp_spin
+        )
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/infer/gen_spin_model_devi.py:228]: RUF059 is in the project's ruff ignore list (pyproject.toml). Matches pattern of all other gen_* scripts.

File: source/tests/infer/gen_spin.py [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+    )
+    print(f"\n// NoPbc total energy: {e_np[0, 0]:.18e}")  # noqa: T201
+    print_cpp_spin_values("NoPbc reference values", ae_np, f_np, fm_np, v_np, av_np)
+
+    # ---- 5. Verify .pth gives same results ----
+    if os.path.exists(pth_path):
+        dp_pth = DeepPot(pth_path)
+        e_pth, f_pth, v_pth, ae_pth, av_pth, fm_pth, _ = dp_pth.eval(
+            coord, box, atype, atomic=True, spin=spin
+        )
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/infer/gen_spin.py:191]: RUF059 is ignored in project ruff config. Code-gen script, not production code.

File: source/tests/pt_expt/infer/test_deep_eval_spin.py [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+    def test_eval_pbc_nonatomic(self, spin_model_files, ext) -> None:
+        """Test PBC evaluation with atomic=False."""
+        from deepmd.infer import (
+            DeepPot,
+        )
+
+        files, ref, _ = spin_model_files
+        dp = DeepPot(files[ext])
+
+        e, f, v, fm, mm = dp.eval(COORD, BOX, ATYPE, atomic=False, spin=SPIN)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/pt_expt/infer/test_deep_eval_spin.py:271]: RUF059 is ignored in project ruff config. Consistent with existing test files.

File: source/api_cc/src/commonPTExpt.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/commonPTExpt.h:431]: Yes. All PyTorch file formats (`.pth`, `.pte`, `.pt2`) always use ZIP STORED (compression method 0, no compression) for every entry. I verified this empirically on our `.pt2` files — every entry has `compress_size == file_size`.

The reason is that PyTorch needs to `mmap` tensor data directly from the archive without decompression. Their C++ writer (`caffe2::serialize::PyTorchStreamWriter`) and the Python `torch.export.save` path both write uncompressed entries with 64-byte alignment for direct memory mapping.

Adding an inline comment to document this assumption is a good idea though.

File: deepmd/pt_expt/utils/serialization.py [hunk 17]
```diff
@@ -16,6 +16,26 @@
...
+    for node in list(graph.nodes):
+        if (
+            node.op == "call_function"
+            and node.target is torch.ops.aten._assert_scalar.default
+        ):
+            graph.erase_node(node)
+    graph.eliminate_dead_code()
+    graph_module.recompile()
+
+
```
Comment (iProzd) [inline deepmd/pt_expt/utils/serialization.py:48]: `_strip_shape_assertions` removes all `aten._assert_scalar` nodes from the exported graph, not just the spin-related `Ne(nall, nloc)` guard. This could silently strip useful shape-consistency checks inserted by torch.export. 
Maybe filter it by the assertion message/args to only remove the known-spurious nall != nloc guard?

File: deepmd/pt_expt/infer/deep_eval.py [hunk 18]
```diff
@@ -124,10 +125,39 @@ def _init_from_model_json(self, model_json_str: str) -> None:
 
         model_dict = json.loads(model_json_str)
         model_dict = _json_to_numpy(model_dict)
-        self._dpmodel = BaseModel.deserialize(model_dict["model"])
+        model_data = model_dict["model"]
+
+        if "spin" in model_data and "backbone_model" in model_data:
+            from deepmd.pt_expt.model.spin_model import (
+                SpinModel,
+            )
```
Comment (iProzd) [inline deepmd/pt_expt/infer/deep_eval.py]: Spin model detection is inconsistent across entry points:
`_init_from_model_json` here uses structural detection, but the `_trace_and_export` in `deepmd/pt_expt/utils/serialization.py` uses type-field detection:
`is_spin = data["model"].get("type") == "spin_ener"`. Could this be unified?

**Review by iProzd — CHANGES_REQUESTED** [review]

File: deepmd/pt_expt/utils/serialization.py (hunk 17 above)
Comment (wanghan-iapcm) [reply deepmd/pt_expt/utils/serialization.py:48]: Good catch — fixed to only remove assertions whose message contains both `nall` and `nloc`, preserving all other shape-consistency checks.

File: deepmd/pt_expt/infer/deep_eval.py (hunk 18 above)
Comment (wanghan-iapcm) [reply deepmd/pt_expt/infer/deep_eval.py]: Unified — `_init_from_model_json` now uses `model_data.get("type") == "spin_ener"`, same as `_trace_and_export`.

File: deepmd/pt_expt/utils/serialization.py (hunk 17 above)
Comment (wanghan-iapcm) [reply deepmd/pt_expt/utils/serialization.py:48]: Update: filtering by assertion message isn't feasible — `torch.export` uses opaque symbolic variable names in the messages (e.g. `Ne(s22, s96)`) rather than human-readable `nall`/`nloc`. Since this function is only called for spin models (guarded by `if is_spin`), and all deferred shape assertions in the spin graph come from the nall/nloc relationship, removing all `_assert_scalar` nodes is safe in this context. Updated the docstring to document this reasoning.

**Review by iProzd — APPROVED** [review]

**Review by OutisLi — APPROVED** [review]
